File: crosshair.cpp
#include "crosshair.h"
#include "./ui_crosshair.h"
#include <math.h>
#include <QtCore>
#include <QPen>
#include <QPainter>
#include<QScreen>

crosshair::crosshair(struct RGBA rgba,
                     QWidget *parent,
                     int window_size,
                     int stroke_width,
                     int distance): QMainWindow(parent), ui(new Ui::crosshair)
{
    // Init checks
    assert(("Distance is too long!", distance <= window_size / 2 - 1));
    assert(("Window size must be an even number!", window_size % 2 == 0));

    ui->setupUi(this);

    this->window_size = window_size;
    this->stroke_width = stroke_width;
    this->rgba = rgba;
    this->distance = distance;
    this->pen = QPen(QColor(rgba.r, rgba.g, rgba.b, rgba.a));
    this->pen.setWidth(stroke_width);

    // Resize window to desired size
    this->resize(window_size, window_size);
    // Disable small resize icon
    this->statusBar()->setSizeGripEnabled(false);

    // Set windows features
    // Transparent, on top, and frameless
    this->setWindowFlags(
        Qt::FramelessWindowHint |
        Qt::WindowStaysOnTopHint |
        Qt::WindowTransparentForInput |
        Qt::Tool);

    // Set transparent attribute
    this->setAttribute(Qt::WA_TranslucentBackground);

    // Center screen
    move(screen()->geometry().center() - frameGeometry().center());
}

crosshair::~crosshair()
{
    delete ui;
}

// Override the paintEvent method
void crosshair::paintEvent(QPaintEvent *e) {

    // Distance from the center
    // (higher the number the closer it is from the center)
    QPainter painter(this);
    painter.setPen(this->pen);

    // Half screen size
    int half_s = reinterpret_cast<int>(this->window_size / 2);
    int m_const = 1;
    int dist = this->distance + ceil(this->pen.width() / 2);

    // Drawing the crosshair

    // Top
    painter.drawLine(half_s, 0, half_s, half_s - dist);
     // Bottom
    painter.drawLine(half_s, half_s + dist, half_s, this->window_size - m_const);
    // Left
    painter.drawLine(0, half_s, half_s - dist, half_s);
    // Right
    painter.drawLine(half_s + dist, half_s, this->window_size - m_const, half_s);
}

File: crosshair.py
import math
from PyQt5 import QtCore, QtGui, QtWidgets
from crosshair_exceptions import DistanceIsTooLong, WinSizeMustBeEven


class Crosshair(QtWidgets.QWidget):
    def __init__(self,
                 parent=None,
                 window_size=24,
                 stroke_width=2,
                 rgba: tuple = (0, 255, 0, 255),
                 distance: int = 4):
        # Init super class
        QtWidgets.QWidget.__init__(self, parent)

        # Assertions
        assert distance <= window_size / 2 - 1, DistanceIsTooLong
        assert window_size % 2 == 0, WinSizeMustBeEven

        self.ws = window_size
        self.resize(window_size + 1, window_size + 1)
        self.distance = distance
        # Assign the pen color and stroke size
        self.pen = QtGui.QPen(QtGui.QColor(*rgba))
        self.pen.setWidth(stroke_width)

        # Set windows features
        # Transparent, on top, and frameless
        self.setWindowFlags(
            QtCore.Qt.FramelessWindowHint |
            QtCore.Qt.WindowStaysOnTopHint |
            QtCore.Qt.WindowTransparentForInput |
            QtCore.Qt.Tool)

        # Set transparent bg attribute
        self.setAttribute(QtCore.Qt.WA_TranslucentBackground)

        self.center()

    def exit(self):
        self.close()

    def center(self):
        """ Move the crosshair to the center of the screen"""
        win_ = QtWidgets.QApplication.desktop().screen().rect().center() - self.rect().center() + QtCore.QPoint(1, 1)
        self.move(win_)

    def paintEvent(self, event):
        """ Override the paintEvent method """
        # Distance from the center
        # (higher the number the closer it is from the center)
        painter = QtGui.QPainter(self)
        painter.setPen(self.pen)

        # Half screen size
        half_s = int(self.ws / 2)
        m_const = 1
        dist = self.distance + math.ceil(self.pen.width() / 2)

        # Drawing the crosshair
        painter.drawLine(half_s, 0, half_s, half_s - dist)  # Top
        painter.drawLine(half_s, half_s + dist, half_s, self.ws - m_const)  # Bottom
        painter.drawLine(0, half_s, half_s - dist, half_s)  # Left
        painter.drawLine(half_s + dist, half_s, self.ws - m_const, half_s)  # Right
